### DISCUSSION TITLE
Calling TypeScript experts!

### DISCUSSION THREAD

**Discussion by djipco** [djipco/webmidi#249] [General]
I really want to provide decent TyoeScript support for the library. However, I'm not a TypeScript developer and it has proven more challenging than I had expected. We currently have an outstanding [issue](https://github.com/djipco/webmidi/issues/247) for which I could really use some help. Any TypeScript experts around these parts?

**Comment by rtman** [comment]
I might be able to help can you link me to the issue and I'll give it a look.

*Edit* Found the link.

**Reply by rtman** [reply]
nvm its fixed already!

**Reply by djipco** [reply]
Thanks for offering to help!

**Comment by leocaseiro** [comment]
Hi there, I've just started to use this library, and typescript can be improved a lot.

I was wondering how I could contribute to it.
I wonder what kind of automation is already there, besides this script https://github.com/djipco/webmidi/blob/master/scripts/typescript-declarations/generate.js.

I've noticed that this project has very good usage of jsdocs. Have you ever considered using the jsdoc as static type?
Ref: https://gils-blog.tayar.org/posts/jsdoc-typings-all-the-benefits-none-of-the-drawbacks/
Ref2: https://github.com/DavidWells/types-with-jsdocs

If you are interested I can help set up, and improve some of the types via jsdoc.  

With that, we no longer would need to manually type all of https://github.com/djipco/webmidi/blob/master/typescript/webmidi.d.ts
I am assuming that is the current scenario.

I am very happy to discuss this on discord or whatever suits you.

**Reply by djipco** [reply]
At first, I was generating the `.d.ts` from the jsdoc comments. However, I ran into a bunch of issues and it turned out to be easier, at least for me, to have a separate file.

Having said that, I'm not a TypeScript expert at all. It would definitely be best if we could generate the `.d.ts` file automatically from the jsdoc comments. If you can take a stab at it, that would be great!

**Reply by djipco** [reply]
I will have more time to discuss this further next week when I'm back hone.

**Reply by leocaseiro** [reply]
Hi @djipco, thank you so much for getting back to me so quick.

I wonder what issues you had in the past, hope we can get pass to it.

I've got ask (sorry), do you have any intention at all to use TypeScript in the future? 

If not, I'm very happy to try fix the issues with JsDoc -> TS

**Comment by djipco** [comment]
> I wonder what issues you had in the past, hope we can get pass to it.

One issue is that TypeScript requires the use of many types that simply are not documented in JSDoc because they are not needed by vanilla JavaScript. Another problem is the name collision between the Web MIDI API types and the library's types. A knowledgeable TypeScript dev might be able to overcome all of these. I don't know. 

One thing that is for sure is that I want to keep the codebase (including jsdoc) free of TypeScript-specific content. So, it may be necessary to isolate TypeScript-specific stuff to its own area. The cleaner, the better. The simpler, the better.

> I've got ask (sorry), do you have any intention at all to use TypeScript in the future?

If you mean use TypeScript to develop the library, the answer is no. However, I want to support TypeScript developers who want to use WEBMIDI.js in a TypeScript context. 

> If not, I'm very happy to try fix the issues with JsDoc -> TS

This is great! You will have my support in doing so!

**Reply by leocaseiro** [reply]
Awesome, thank you for clarifying that. I will look into that soon.
